libraries/chain/wasm_interface.cpp

```diff
@@ -93,7 +93,15 @@ namespace eosio { namespace chain {
       if(my->eosvmoc) {
          const chain::eosvmoc::code_descriptor* cd = nullptr;
          try {
-            cd = my->eosvmoc->cc.get_descriptor_for_code(code_hash, vm_version, context.control.is_write_window());
+            eosio::chain::eosvmoc::get_cd_failure failure = eosio::chain::eosvmoc::get_cd_failure::temporary;
+            cd = my->eosvmoc->cc.get_descriptor_for_code(code_hash, vm_version, context.control.is_write_window(), failure);
+            if (!cd && context.trx_context.is_read_only()) {
+               if (failure == eosio::chain::eosvmoc::get_cd_failure::temporary) {
+                  EOS_ASSERT(false, ro_trx_temporary, "get_descriptor_for_code failed with temporary failure");
+               } else {
+                  EOS_ASSERT(false, ro_trx_permanent, "get_descriptor_for_code failed with permanent failure");
+               }
+            }
          }
          catch(...) {
```

Comment:
This is not quite right as the `catch(...)` catches these `EOS_ASSERT`. This should be more like:
```
index 4cbdfdbadf..5cd75786f0 100644
--- a/libraries/chain/wasm_interface.cpp
+++ b/libraries/chain/wasm_interface.cpp
@@ -92,16 +92,9 @@ namespace eosio { namespace chain {
 #ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
       if(my->eosvmoc) {
          const chain::eosvmoc::code_descriptor* cd = nullptr;
+         eosio::chain::eosvmoc::get_cd_failure failure = eosio::chain::eosvmoc::get_cd_failure::temporary;
          try {
-            eosio::chain::eosvmoc::get_cd_failure failure = eosio::chain::eosvmoc::get_cd_failure::temporary;
             cd = my->eosvmoc->cc.get_descriptor_for_code(code_hash, vm_version, context.control.is_write_window(), failure);
-            if (!cd && context.trx_context.is_read_only()) {
-               if (failure == eosio::chain::eosvmoc::get_cd_failure::temporary) {
-                  EOS_ASSERT(false, ro_trx_temporary, "get_descriptor_for_code failed with temporary failure");
-               } else {
-                  EOS_ASSERT(false, ro_trx_permanent, "get_descriptor_for_code failed with permanent failure");
-               }
-            }
          }
          catch(...) {
             //swallow errors here, if EOS VM OC has gone in to the weeds we shouldn't bail: continue to try and run baseline
@@ -114,6 +107,12 @@ namespace eosio { namespace chain {
          if(cd) {
             my->eosvmoc->exec->execute(*cd, my->eosvmoc->mem, context);
             return;
+         } else if (context.trx_context.is_read_only()) {
+            if (failure == eosio::chain::eosvmoc::get_cd_failure::temporary) {
+               EOS_ASSERT(false, ro_trx_temporary, "get_descriptor_for_code failed with temporary failure");
+            } else {
+               EOS_ASSERT(false, ro_trx_permanent, "get_descriptor_for_code failed with permanent failure");
+            }
          }
       }
 #endif
```

Reply:
Thanks for raising the issue and detailed solution.